subsys/net/ip/l2/openthread/openthread.c

```diff
@@ -0,0 +1,348 @@
+		if (!pkt_buf) {
+			SYS_LOG_ERR("Failed to get fragment buf");
+			net_pkt_unref(pkt);
+			otMessageFree(aMessage);
+			return;
+		}
+
+		read_len = otMessageRead(aMessage, offset,
+			pkt_buf->data, net_buf_tailroom(pkt_buf));
+
+		if (!read_len) {
+			net_buf_unref(pkt_buf);
+			break;
+		}
+
+		net_buf_add(pkt_buf, read_len);
+
+		if (!prev_buf) {
+			net_pkt_frag_insert(pkt, pkt_buf);
+		} else {
+			net_buf_frag_insert(prev_buf, pkt_buf);
+		}
+
+		prev_buf = pkt_buf;
+
+		offset += read_len;
+	}
+
+	otMessageFree(aMessage);
+
+	SYS_LOG_DBG("Injecting Ip6 packet to Zephyr net stack");
+
+#if defined(CONFIG_OPENTHREAD_L2_DEBUG_DUMP_IPV6)
+	dump_pkt("Received IPv6 packet", pkt);
+#endif
+
+	if (!pkt_list_is_full(ot_context)) {
+		if (net_recv_data(ot_context->iface, pkt) < 0) {
+			SYS_LOG_ERR("net_recv_data failed");
+			net_pkt_unref(pkt);
+			return;
+		}
+
+		pkt_list_add(ot_context, pkt);
+	} else {
+		SYS_LOG_INF("Pacet list is full");
+		net_pkt_unref(pkt);
+	}
+}
+
+static void openthread_process(void *context, void *arg2, void *arg3)
+{
+	struct openthread_context *ot_context = context;
+
+	while (1) {
+		while (otTaskletsArePending(ot_context->instance)) {
+			otTaskletsProcess(ot_context->instance);
+		}
+		PlatformProcessDrivers(ot_context->instance);
+
+		k_sem_take(&ot_sem, K_FOREVER);
+	}
+}
+
+static enum net_verdict openthread_recv(struct net_if *iface,
+	struct net_pkt *pkt)
+{
+	struct openthread_context *ot_context = net_if_l2_data(iface);
+
+	if (pkt_list_peek(ot_context) == pkt) {
+		pkt_list_remove_last(ot_context);
+		SYS_LOG_DBG("Got injected Ip6 packet, sending to upper layers");
+#if defined(CONFIG_OPENTHREAD_L2_DEBUG_DUMP_IPV6)
+		dump_pkt("Injected IPv6 packet", pkt);
+#endif
+		return NET_CONTINUE;
+	}
+
+	SYS_LOG_DBG("Got 802.15.4 packet, sending to OT");
+
+	otRadioFrame recv_frame;
+
+	recv_frame.mPsdu = net_buf_frag_last(pkt->frags)->data;
+	/* Length inc. CRC. */
+	recv_frame.mLength = net_buf_frags_len(pkt->frags);
+	/* TODO: get channel from packet */
+	recv_frame.mChannel = CONFIG_OPENTHREAD_CHANNEL;
+	recv_frame.mLqi = net_pkt_ieee802154_lqi(pkt);
+	recv_frame.mPower = net_pkt_ieee802154_rssi(pkt);
+
+#if defined(CONFIG_OPENTHREAD_L2_DEBUG_DUMP_15_4)
+	dump_pkt("Received 802.15.4 frame", pkt);
+#endif
+
+#if OPENTHREAD_ENABLE_DIAG
+	if (otPlatDiagModeGet()) {
+		otPlatDiagRadioReceiveDone(ot_context->instance,
+			&recv_frame, OT_ERROR_NONE);
+	} else
+#endif
+	{
+		otPlatRadioReceiveDone(ot_context->instance,
+			&recv_frame, OT_ERROR_NONE);
+	}
+
+	net_pkt_unref(pkt);
+
+	return NET_OK;
+}
+
+enum net_verdict openthread_send(struct net_if *iface, struct net_pkt *pkt)
+{
+	struct openthread_context *ot_context = net_if_l2_data(iface);
+	enum net_verdict ret = NET_OK;
+	otMessage *message;
+
+	SYS_LOG_DBG("Sending Ip6 packet to ot stack");
+
+	message = otIp6NewMessage(ot_context->instance, true);
+	if (message == NULL) {
+		goto exit;
+	}
+
+	struct net_buf *frag;
```

Comment:
variable declaration is always at the beginning of a code block, so move this above.